### PR TITLE
Implement Training/Validation Data Split Strategy (Issue 49)

### PR DESCRIPTION
**Pull request by mlaurako**
## Summary
This PR implements a proper training/validation dataset split strategy as outlined in **Issue #49** and adds an additional checkpoint resume option to support different workflow preferences.

### **New Training/Validation Data Strategy**
- **Problem Solved**: Previously, the model randomly selected a 14-day window and performed train/val split within that window, leading to potential data leakage and inconsistent validation sets.
- **Solution**: Implemented temporal train/validation split where:
  - Training data is retrieved during the `on_train_epoch_start` callback
  - Validation data is retrieved during the `on_validation_epoch_start` callback  
  - Ensures proper temporal separation between training and validation sets
  - Follows general ML best practices for time series data splitting
```python
class ResampleDataCallback(pl.Callback):
    def on_train_epoch_start(self, trainer, pl_module):
        """Sample from training date range"""
    def on_validation_epoch_start(self, trainer, pl_module):
        """Sample from validation date range"""
```
### **Additional Checkpoint Resume Method (`--resume_from_latest`)**
This feature adds automatic detection of the most recent `last-v${version_number}.ckpt` checkpoint.
- **Available Methods**:
  1. **Existing method**: `--resume_from_checkpoint checkpoints/last-v5.ckpt`
  2. **Additional method**: `--resume_from_latest` (automatically finds newest checkpoint)
```bash
# Method 1: Specify exact checkpoint (existing)
srun --export=ALL --kill-on-bad-exit=1 --cpu-bind=cores python train_gnn.py --resume_from_checkpoint checkpoints/last-v5.ckpt
# Method 2: Auto-detect latest checkpoint (additional)  
srun --export=ALL --kill-on-bad-exit=1 --cpu-bind=cores python train_gnn.py --resume_from_latest
```
You may validate the checkpoint version by looking for the following message in the output file:
```
[INFO] Auto-resuming from: checkpoints/last-v10.ckpt
```

## Results & Validation
Results are **compatible with the previous implementation**, confirming no performance degradation.
- **File Locations**: 
    `/scratch3/NCEPDEV/da/Mu-Chieh.Ko/OCELOT/DEV/train-val2/ocelot/gnn_model/fig_evaluation_100epoch`
    `/scratch3/NCEPDEV/da/Mu-Chieh.Ko/OCELOT/DEV/train-val2/ocelot/gnn_model/val_csv_100epoch_inv1`
- **Validation**: 100-epoch training completed successfully with expected prediction patterns

### **Known Limitations (Future Work)**
The runtime has increased since we now need to run `fit` twice to create datamodules in one epoch. Additionally, each mini-epoch trains on only 14 days of data. My initial thought was to decrease validation frequency from every epoch to every 5 epochs (running validation once after all data has been seen by the model). 
- **Validation Frequency**: Attempted to implement `check_val_every_n_epoch = 5` but encountered LR scheduler conflicts:
  ```
  MisconfigurationException: ReduceLROnPlateau conditioned on metric val_loss which is not available. 
  Available metrics are: ['train_loss', 'train_loss_step', 'rollout_steps', 'train_loss_epoch']
  ```
- Both direct trainer kwargs and callback-based approaches failed
- Requires further investigation to shorten training runtime

## Future Work
1. **Validation Frequency Control**: Resolve LR scheduler conflicts to enable `check_val_every_n_epoch > 1`
2. **Performance Optimization**: Investigate runtime improvements for longer training periods

### CODE REVIEW

**Review by xincjin-NOAA — APPROVED** [review]
looks good to me

File: gnn_model/train_gnn.py [hunk 1]
```diff
@@ -71,13 +77,29 @@ def main():
     # --- DEFINE THE FULL DATE RANGE FOR THE EXPERIMENT ---
     FULL_START_DATE = "2024-04-01"
     FULL_END_DATE = "2024-07-01"  # e.g., 3 months of data
-    WINDOW_DAYS = 7  # The size of the window for each epoch
+    TRAIN_WINDOW_DAYS = 14  # The size of the training window for each epoch
+    VALID_WINDOW_DAYS = 3   # The size of the validation window for each epoch
```
Comment (azadeh-gh) [inline gnn_model/train_gnn.py:82]: Tested the modified training/validation data split strategy. With the longer 14-day training window, runtime roughly doubled. Changing the setting to a 7-day training window with 1-day validation brings the runtime back in line with the original version (reducing from ~13 hours to ~7 hours).
Note: we should standardize on the following configuration for efficient runs:
TRAIN_WINDOW_DAYS = 7   # Training window size per epoch  
VALID_WINDOW_DAYS = 1   # Validation window size per epoch
This setup still produces good results while keeping runtime manageable.

**Review by azadeh-gh — APPROVED** [review]
Tested the modified training/validation data split strategy. With the longer 14-day training window, runtime roughly doubled. Changing the setting to a 7-day training window with 1-day validation brings the runtime back in line with the original version (reducing from ~13 hours to ~7 hours).

The changes look good to me, and I approve and will merge this PR into main.

**Comment by azadeh-gh** [comment]
Tested the modified training/validation data split strategy. With the longer 14-day training window, runtime roughly doubled. Changing the setting to a 7-day training window with 1-day validation brings the runtime back in line with the original version (reducing from ~13 hours to ~7 hours).

The changes look good to me, and I approve and will merge this PR into main.

Note: we should standardize on the following configuration for efficient runs:

TRAIN_WINDOW_DAYS = 7   # Training window size per epoch  
VALID_WINDOW_DAYS = 1   # Validation window size per epoch


This setup still produces good results while keeping runtime manageable.
